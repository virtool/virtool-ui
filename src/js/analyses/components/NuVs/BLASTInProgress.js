import React from "react";
import { formatDistanceStrict, addSeconds } from "date-fns";
import styled from "styled-components";
import { Box, ExternalLink, Icon, Loader, RelativeTime } from "../../../base";

const ridRoot =
    "https://blast.ncbi.nlm.nih.gov/Blast.cgi?\
    CMD=Web&PAGE_TYPE=BlastFormatting&OLD_BLAST=false&GET_RID_INFO=on&RID=";

export const RIDLink = ({ rid }) => {
    if (rid) {
        return (
            <span>
                <span> with RID </span>
                <ExternalLink href={ridRoot + rid}>
                    {rid}{" "}
                    <sup>
                        <Icon name="new-tab" />
                    </sup>
                </ExternalLink>
            </span>
        );
    }

    return null;
};

const StyledRIDTiming = styled.div`
    font-size: ${props => props.theme.fontSize.sm};
    margin-left: auto;
`;

export const RIDTiming = ({ interval, lastCheckedAt }) => {
    if (lastCheckedAt) {
        const nextCheckAt = addSeconds(new Date(lastCheckedAt), interval);
        const relativeNext = formatDistanceStrict(new Date(nextCheckAt), Date.now());

        return (
            <StyledRIDTiming>
                Last checked <RelativeTime time={lastCheckedAt} />. Checking again in {relativeNext}
            </StyledRIDTiming>
        );
    }

    return null;
};

const StyledBLASTInProgress = styled(Box)`
    align-items: flex-start;
    display: flex;

    div:first-of-type {
        margin-right: 5px;
    }
`;

export const BLASTInProgress = ({ interval, lastCheckedAt, rid }) => {
    return (
        <StyledBLASTInProgress>
            <Loader size="16px" color="primary" />
            <div>
                <div>
                    <span>BLAST in progress</span>
                    <RIDLink rid={rid} />
                </div>
                <RIDTiming interval={interval} lastCheckedAt={lastCheckedAt} />
            </div>
        </StyledBLASTInProgress>
    );
};
