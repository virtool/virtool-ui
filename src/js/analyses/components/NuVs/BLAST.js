import { map } from "lodash-es";
import numbro from "numbro";
import React, { useCallback } from "react";
import { connect } from "react-redux";
import styled from "styled-components";
import { Alert, Box, BoxGroup, BoxGroupHeader, BoxTitle, Button, ExternalLink, Icon, Table } from "../../../base";

import { blastNuvs } from "../../actions";
import { getActiveHit } from "../../selectors";
import { BLASTError } from "./BLASTError";
import { BLASTInProgress } from "./BLASTInProgress";

export const BLASTButton = styled(Button)`
    font-weight: 600;
    margin-left: auto;
`;

const StyledBLASTResultsHeader = styled(BoxGroupHeader)`
    display: flex;
    flex-direction: row;
    justify-content: space-between;
`;

export const BLASTResults = ({ hits, onBlast }) => {
    const components = map(hits, (hit, index) => (
        <tr key={index}>
            <td>
                <ExternalLink href={`https://www.ncbi.nlm.nih.gov/nuccore/${hit.accession}`}>
                    {hit.accession}
                </ExternalLink>
            </td>
            <td>{hit.name}</td>
            <td>{hit.evalue}</td>
            <td>{hit.score}</td>
            <td>{numbro(hit.identity / hit.align_len).format("0.00")}</td>
        </tr>
    ));

    return (
        <BoxGroup>
            <StyledBLASTResultsHeader>
                <strong>NCBI BLAST</strong>
                <a href="#" onClick={onBlast}>
                    <Icon name="redo" /> Retry
                </a>
            </StyledBLASTResultsHeader>
            <Table>
                <thead>
                    <tr>
                        <th>Accession</th>
                        <th>Name</th>
                        <th>E-value</th>
                        <th>Score</th>
                        <th>Identity</th>
                    </tr>
                </thead>
                <tbody>{components}</tbody>
            </Table>
        </BoxGroup>
    );
};

export const NuVsBLAST = ({ analysisId, blast, sequenceIndex, onBlast }) => {
    const handleBlast = useCallback(() => onBlast(analysisId, sequenceIndex), [analysisId, sequenceIndex]);

    if (blast) {
        if (blast.error) {
            return <BLASTError error={blast.error} onBlast={handleBlast} />;
        }

        if (blast.ready) {
            if (blast.result.hits.length) {
                return <BLASTResults hits={blast.result.hits} onBlast={handleBlast} />;
            }

            return (
                <Box>
                    <BoxTitle>NCBI BLAST</BoxTitle>
                    <p>No BLAST hits found.</p>
                </Box>
            );
        }

        return <BLASTInProgress interval={blast.interval} lastCheckedAt={blast.last_checked_at} rid={blast.rid} />;
    }

    return (
        <Alert color="purple" level>
            <Icon name="info-circle" />
            <span>This sequence has no BLAST information attached to it.</span>
            <BLASTButton color="purple" icon="cloud" onClick={handleBlast}>
                BLAST at NCBI
            </BLASTButton>
        </Alert>
    );
};

export const mapStateToProps = state => {
    const { blast, index } = getActiveHit(state);

    return {
        analysisId: state.analyses.detail.id,
        blast,
        sequenceIndex: index
    };
};

export const mapDispatchToProps = dispatch => ({
    onBlast: (analysisId, sequenceIndex) => {
        dispatch(blastNuvs(analysisId, sequenceIndex));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(NuVsBLAST);
