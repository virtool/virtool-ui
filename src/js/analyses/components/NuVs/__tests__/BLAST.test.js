import { BLASTButton, BLASTResults, NuVsBLAST } from "../BLAST";
import { BLASTInProgress } from "../BLASTInProgress";

describe("<BLASTInProgress />", () => {
    it.each(["ABC123", null])("should render when [rid=%p]", rid => {
        const wrapper = shallow(
            <BLASTInProgress interval={5} lastCheckedAt={"2018-02-14T17:12:00.000000Z"} rid={rid} />
        );
        expect(wrapper).toMatchSnapshot();
    });
});

describe("<BLASTResults />", () => {
    it("should render", () => {
        const props = {
            hits: [
                {
                    name: "test",
                    accession: "NC123",
                    evalue: 3,
                    score: 1,
                    identity: 2,
                    align_len: 4
                }
            ]
        };
        const wrapper = shallow(<BLASTResults {...props} />);
        expect(wrapper).toMatchSnapshot();
    });
});

describe("<NuVsBLAST />", () => {
    let props;

    beforeEach(() => {
        props = {
            blast: {
                ready: true,
                result: {
                    hits: [
                        {
                            name: "bar",
                            accession: "BAR123",
                            evalue: 3,
                            score: 1,
                            identity: 2,
                            align_len: 4
                        }
                    ]
                }
            },
            analysisId: "foo",
            sequenceIndex: 5,
            onBlast: jest.fn()
        };
    });

    it("should render", () => {
        const wrapper = shallow(<NuVsBLAST {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render with no BLAST hits", () => {
        props.blast.result.hits = [];
        const wrapper = shallow(<NuVsBLAST {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render with an in-progress BLAST", () => {
        props.blast = {
            ready: false,
            interval: 5,
            last_checked_at: "2018-02-14T17:12:00.000000Z",
            rid: "ABC123"
        };
        const wrapper = shallow(<NuVsBLAST {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should render alert when no blast data available", () => {
        props.blast = null;
        const wrapper = shallow(<NuVsBLAST {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("should call blastNuvs() when BLAST button clicked", () => {
        props.blast = null;
        const wrapper = shallow(<NuVsBLAST {...props} />);
        wrapper.find(BLASTButton).simulate("click");
        expect(props.onBlast).toHaveBeenCalledWith(props.analysisId, props.sequenceIndex);
    });
});
