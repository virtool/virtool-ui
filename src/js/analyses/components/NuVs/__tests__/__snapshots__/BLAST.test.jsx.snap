// Vitest Snapshot v1

exports[`<BLASTInProgress /> > should render when [rid=%p] ABC123 1`] = `
<Styled(Box)>
  <Loader
    color="primary"
    size="16px"
  />
  <div>
    <div>
      <span>
        BLAST in progress
      </span>
      <RIDLink
        rid="ABC123"
      />
    </div>
    <RIDTiming
      interval={5}
      lastCheckedAt="2018-02-14T17:12:00.000000Z"
    />
  </div>
</Styled(Box)>
`;

exports[`<BLASTInProgress /> > should render when [rid=%p] null 1`] = `
<Styled(Box)>
  <Loader
    color="primary"
    size="16px"
  />
  <div>
    <div>
      <span>
        BLAST in progress
      </span>
      <RIDLink
        rid={null}
      />
    </div>
    <RIDTiming
      interval={5}
      lastCheckedAt="2018-02-14T17:12:00.000000Z"
    />
  </div>
</Styled(Box)>
`;

exports[`<BLASTResults /> > should render 1`] = `
<BoxGroup>
  <Styled(BoxGroupHeader)>
    <strong>
      NCBI BLAST
    </strong>
    <Button
      color="blue"
    >
      <Icon
        name="redo"
      />
      Retry
    </Button>
  </Styled(BoxGroupHeader)>
  <Table>
    <thead>
      <tr>
        <th>
          Accession
        </th>
        <th>
          Name
        </th>
        <th>
          E-value
        </th>
        <th>
          Score
        </th>
        <th>
          Identity
        </th>
      </tr>
    </thead>
    <tbody>
      <tr
        key="0"
      >
        <td>
          <ExternalLink
            href="https://www.ncbi.nlm.nih.gov/nuccore/NC123"
          >
            NC123
          </ExternalLink>
        </td>
        <td>
          test
        </td>
        <td>
          3
        </td>
        <td>
          1
        </td>
        <td>
          0.50
        </td>
      </tr>
    </tbody>
  </Table>
</BoxGroup>
`;

exports[`<NuVsBLAST /> > should render 1`] = `
<BLASTResults
  hits={
    [
      {
        "accession": "BAR123",
        "align_len": 4,
        "evalue": 3,
        "identity": 2,
        "name": "bar",
        "score": 1,
      },
    ]
  }
  onBlast={[Function]}
/>
`;

exports[`<NuVsBLAST /> > should render alert when no blast data available 1`] = `
<Alert
  color="purple"
  level={true}
>
  <Icon
    name="info-circle"
  />
  <span>
    This sequence has no BLAST information attached to it.
  </span>
  <Styled(Button)
    color="purple"
    icon="cloud"
    onClick={[Function]}
  >
    BLAST at NCBI
  </Styled(Button)>
</Alert>
`;

exports[`<NuVsBLAST /> > should render with an in-progress BLAST 1`] = `
<BLASTInProgress
  interval={5}
  lastCheckedAt="2018-02-14T17:12:00.000000Z"
  rid="ABC123"
/>
`;

exports[`<NuVsBLAST /> > should render with no BLAST hits 1`] = `
<Box>
  <BoxTitle>
    NCBI BLAST
  </BoxTitle>
  <p>
    No BLAST hits found.
  </p>
</Box>
`;
